wgengine/netstack/netstack.go

```diff
@@ -37,73 +38,39 @@ import (
 	"tailscale.com/wgengine/tstun"
 )
 
-func Impl(logf logger.Logf, tundev *tstun.TUN, e wgengine.Engine, mc *magicsock.Conn) error {
+type Impl struct {
+	ipstack *stack.Stack
+	linkEP  *channel.Endpoint
+	tundev  *tstun.TUN
+	e       wgengine.Engine
+	mc      *magicsock.Conn
+	logf    logger.Logf
+}
+
+const nicID = 1
+
+func Create(logf logger.Logf, tundev *tstun.TUN, e wgengine.Engine, mc *magicsock.Conn) (wgengine.FakeImpl, error) {
 	if mc == nil {
-		return errors.New("nil magicsock.Conn")
+		return nil, errors.New("nil magicsock.Conn")
 	}
 	if tundev == nil {
-		return errors.New("nil tundev")
+		return nil, errors.New("nil tundev")
 	}
 	if logf == nil {
-		return errors.New("nil logger")
+		return nil, errors.New("nil logger")
 	}
 	if e == nil {
-		return errors.New("nil Engine")
+		return nil, errors.New("nil Engine")
 	}
 	ipstack := stack.New(stack.Options{
 		NetworkProtocols:   []stack.NetworkProtocolFactory{ipv4.NewProtocol},
 		TransportProtocols: []stack.TransportProtocolFactory{tcp.NewProtocol, udp.NewProtocol, icmp.NewProtocol4},
 	})
-
 	const mtu = 1500
 	linkEP := channel.New(512, mtu, "")
-
-	const nicID = 1
 	if err := ipstack.CreateNIC(nicID, linkEP); err != nil {
-		log.Fatal(err)
+		return nil, errors.New("could not create netstack NIC: " + err.String())
```

Comment:
`return nil, fmt.Errorf("could not create netstack NIC: %w", err)`

The `%w` wraps the error, so its semantic properties aren't lost for upper layers to query. (Whereas String flattens it into a meaningless error that's just a string)

Reply:
That doesn't seem to work in this case, because netstack is using their own error type here, which doesn't implement the standard error interface: https://github.com/google/netstack/blob/master/tcpip/tcpip.go#L52

Reply:
That works **even more**. Use `%w` so that underlying error of theirs doesn't get lost.

Reply:
Huh. gopls was complaining that the type was wrong, but it seems to compile fine. Guess I'll make it so then.

Reply:
Oh, ugh, they did:

```
CreateNIC(id tcpip.NICID, ep LinkEndpoint) *tcpip.Error
```

That's not idiomatic Go. (See https://golang.org/doc/faq#nil_error)

So, yeah, `gopls` is properly warning you the `faq#nil_error` footgun. In this case it's fine, but to make it obviously safe and clear for future readers, I'd write:

```go
    if tcpipErr := ipstack.CreateNIC(nicID, linkEP); tcpipErr != nil {
             var err error = tcpipErr // tcpipErr is *tcpip.Error
             return nil, fmt.Errorf("could not create netstack NIC: %w", err)
    }
```

Reply:
That second line doesn't seem to work, since again `*tcpip.Error` doesn't have an `Error()` method.